### ISSUE TITLE
[Question] Shape mismatch when concatenating embeddings for sequential ranking model

### ISSUE THREAD

**Issue by datasciyj** [tensorflow/recommenders#618]
Hi,

I'm trying to create ranking model with sequential data. However, my code keeps crashing in model fitting part because of the shape mismatch between embeddings from query model and candidate model. Here is the error message. I think the code crashed when ranking model concatenates embeddings from the query model and candidate model before sending it to the dense layer. The sequence has the 15 purchase history with '0' as a padding.  Any suggestion to solve this would be appreciated.

"Exception encountered when calling layer 'ranking_model_9' (type RankingModel).\n\n{{function_node __wrapped__ConcatV2_N_2_device_/job:localhost/replica:0/task:0/device:CPU:0}} ConcatOp : Ranks of all input tensors should match: shape[0] = [512,32] vs. shape[1] = [512,1,32] [Op:ConcatV2] name: concat\n\nCall arguments received by layer 'ranking_model_9' (type RankingModel):\n  • inputs=('tf.Tensor(shape=(512, 15), dtype=string)', 'tf.Tensor(shape=(512, 1), dtype=string)')",

The model is as below. 
```python
embedding_dimension = 32

query_model = tf.keras.Sequential([
    tf.keras.layers.StringLookup(
      vocabulary=unique_product_ids, mask_token=None),
    tf.keras.layers.Embedding(len(unique_product_ids) + 1, embedding_dimension), 
    tf.keras.layers.GRU(embedding_dimension)
    # tf.keras.layers.LSTM(embedding_dimension)

])

candidate_model = tf.keras.Sequential([
  tf.keras.layers.StringLookup(
      vocabulary=unique_product_ids, mask_token=None),
  tf.keras.layers.Embedding(len(unique_product_ids) + 1, embedding_dimension)
])

class RankingModel(tf.keras.Model):

  def __init__(self):
    super().__init__()
    embedding_dimension = 32

    self._query_model = query_model
    self._candidate_model = candidate_model

    # Compute predictions.
    self.prob = tf.keras.Sequential([
      # Learn multiple dense layers.
      tf.keras.layers.Dense(256, activation="relu"),
      tf.keras.layers.Dense(64, activation="relu"),
      # Make probability predictions in the final layer.
      tf.keras.layers.Dense(1, activation='sigmoid')
  ])

  def call(self, inputs):

    purchase_history, candidates_retrieval = inputs

    query_embedding = self._query_model(purchase_history)
    candidate_embedding = self._candidate_model(candidates_retrieval)

    return self.prob(tf.concat([query_embedding, candidate_embedding], axis=1))

class NextitemModel(tfrs.models.Model):

  def __init__(self):
    super().__init__()
    self.ranking_model: tf.keras.Model = RankingModel()
    self.task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
    loss = tf.keras.losses.BinaryCrossentropy(),
    metrics=[
      tf.keras.metrics.AUC(name='auc'),
      tf.keras.metrics.BinaryAccuracy(name="accuracy"),
      ]
)

  def call(self, features: Dict[str, tf.Tensor]) -> tf.Tensor:
    return self.ranking_model(
        (features["product"], features["target"]))

  def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:
    labels = features.pop("purchase")
    prob_predictions = self(features)

    # The task computes the loss and the metrics.
    return self.task(labels=labels, predictions=prob_predictions)

rankmodel = NextitemModel()
rankmodel.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.1),run_eagerly=True)

rankmodel.fit(train_dataset, epochs=3, verbose=2)
```

**Comment by patrickorlando** [comment]
Hi @datasciyj,

It looks like you have a tensorshape mismatch in your concatenation step.
```
 [512,32] vs. shape[1] = [512,1,32]
 ```

The sequence embedding coming out of your GRU model still has the sequence dimension. You need to squeeze or reshape the tensor to from `(None, 1, 32)` to `(None, 32)`.


Also a tip for formatting your issue. Your code was a little hard to read. I'd suggest you include your code in a code block with syntax highlighting, using the following format.

\```python
YOUR CODE HERE
\```

best of luck

**Comment by datasciyj** [comment]
Hi @patrickorlando, thank you for your comment!

I found [512,1,32] is actually from the candidate embedding not from the sequence embedding.
So my data used to look like 
{'sequence': [[b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'990365809', b'631'], [b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'203', b'11', b'245'] ... , 'target': [[b'800],[b'23'], ...] , 'label':[[1],[1], ... ] }
This data format was same as the sequence data in the tutorial for sequential retrieval model, but the same format didn't work for the ranking model when concatenating those two embeddings. 
So after I figured out the candidate embedding has additional dimension, I ended up removing brackets for 'target' which is candidate data from [[b'800],[b'23'], ...] to [b'800', b'23' .. ], and the model fitting ran without error. 
I'm curious why the same data format doesn't work for the ranking model with sequence data and also if removing bracket from candidate embedding is correct way to solve this problem.

**Comment by patrickorlando** [comment]
@datasciyj, nice work finding a solution.
There is no concatenation in the retrieval task, only matrix multiplication, which is valid even with the extra dimension.

**Comment by patrickorlando** [comment]
Great work finding the root cause @datasciyj!
I've followed the sequential tutorial, the output of the candidate model has shape `(512,32)`, which doesn't match the problem you've described here.

The reason you don't get an error is because there is not concatenation in the retrieval task, only matrix multiplication, which will be valid even with the extra dimension.
`(512, 32) x (32, 1, 512) => (512, 1, 512)`

However this might cause issues when calculating the softmax loss. So it's fine to remove those extra brackets, I would just make sure the shape of my query and candidate vectors is what you expect in both models.

**Comment by datasciyj** [comment]
Thank you @patrickorlando for the detailed explanation! I got why different dimensions was not an issue for the sequential retrieval model tutorial.
I got another question, so I'll open another one. Thanks again.
